### PR TITLE
Add endpoints and merge database

### PR DESCRIPTION
**Pull request by Laursen79**

### CODE REVIEW

**Review by AOsnes — DISMISSED** [review]

File: client/src/server.rs [hunk 1]
```diff
@@ -23,6 +25,21 @@ enum ReqType {
...
+impl Display for ReqType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                ReqType::Get => "GET",
+                ReqType::Post(value) => "POST",
+                ReqType::Put(value) => "PUT",
+                ReqType::Delete(value) => "DELETE",
```
Comment (Darkros1245) [inline client/src/server.rs:37]: Use _ instead of value

File: server/database/init.sql [hunk 2]
```diff
@@ -73,7 +73,7 @@ CREATE TABLE one_time_pre_key_store (
 INSERT INTO
     accounts (aci, pni, auth_token, identity_key)
 VALUES
-    ('bob', NULL, '1236854bff0ad5aa206f924c9c2ff800681f69df4f6963976f144c1842c2ff1b', '\x424f424b4559'::bytea); -- auth_token = BBBBBB
+    ('bob', NULL, '1236854bff0ad5aa206f924c9c2ff800681f69df4f6963976f144c1842c2ff1b', '\x053b258970748dff667800108898e390a1dfabd8fa66889748d903c12fbc5c732d'::bytea); -- auth_token = BBBBBB
```
Comment (Darkros1245) [inline server/database/init.sql:76]: We should also change this for alice

File: server/src/in_memory_db.rs [hunk 3]
```diff
@@ -1,17 +1,254 @@
...
-    pub mailbox: HashMap<String, Vec<Envelope>>,
+#[derive(Clone)]
+pub struct InMemorySignalDatabase {
+    pub mail_queues: Arc<Mutex<HashMap<Device, VecDeque<Envelope>>>>,
+    pub devices: Arc<Mutex<HashMap<ServiceId, Vec<Device>>>>,
+    pub keys:
+        Arc<Mutex<HashMap<ServiceId, HashMap<DeviceId, HashMap<PreKey, Vec<UploadSignedPreKey>>>>>>,
+    pub accounts: Arc<Mutex<HashMap<String, Account>>>,
 }
+/*
```
Comment (Darkros1245) [inline server/src/in_memory_db.rs:33]: Remove comment

File: server/src/in_memory_db.rs [hunk 4]
```diff
@@ -1,17 +1,254 @@
...
+        otpks: Vec<UploadSignedPreKey>,
+        d_owner: &Device,
+        a_owner: &Account,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    async fn get_one_time_pre_key_count(&self, account: &Account) -> Result<u32> {
+        todo!()
+        /*
```
Comment (Darkros1245) [inline server/src/in_memory_db.rs:196]: Remove or add comment

File: server/src/main.rs [hunk 5]
```diff
@@ -1,6 +1,8 @@
-mod create_server;
-mod db;
-mod in_memory_db;
+#![allow(unused)]
```
Comment (Darkros1245) [inline server/src/main.rs:1]: Are we going to need this later?

File: server/src/postgres.rs [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+        new_aci: Option<String>,
+    ) -> Result<()> {
+        sqlx::query!(
+            r#"
+        UPDATE
+            accounts
+        SET
+            aci = $2
+        WHERE
+            aci = $1
```
Comment (Darkros1245) [inline server/src/postgres.rs:91]: Add COALESCE

File: server/src/postgres.rs [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+        new_pni: Option<String>,
+    ) -> Result<()> {
+        sqlx::query!(
+            r#"
+        UPDATE
+            accounts
+        SET
+            pni = $2
+        WHERE
+            pni = $1
```
Comment (Darkros1245) [inline server/src/postgres.rs:114]: Add COALESCE

File: server/src/postgres.rs [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+    }
+
+    async fn delete_account(&self, aci: Option<String>, pni: Option<String>) -> Result<()> {
+        sqlx::query!(
+            r#"
+        DELETE FROM
+            accounts
+        WHERE
+            aci = $1
+            AND pni = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:132]: Add COALESCE

File: server/src/postgres.rs [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+            devices
+        WHERE
+            device_id = $1
+            AND owner = (
+                SELECT
+                    id
+                FROM
+                    accounts
+                WHERE
+                    aci = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:263]: Add COALESCE

File: server/src/postgres.rs [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+            INNER JOIN devices on devices.id = msq_queue.d_receiver
+        WHERE
+            msq_queue.a_receiver = (
+                SELECT
+                    id
+                FROM
+                    accounts
+                WHERE
+                    aci = $1
+                    AND pni = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:298]: Add COALESCE

File: server/src/postgres.rs [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+            INNER JOIN devices ON devices.id = device_keys.owner
+        WHERE
+            devices.owner = (
+                SELECT
+                    id
+                FROM
+                    accounts
+                WHERE
+                    aci = $1
+                    AND pni = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:439]: Add COALESCE

File: server/src/postgres.rs [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+                INSERT INTO
+                    one_time_pre_key_store (owner, key_id, public_key, signature)
+                SELECT
+                    devices.id, $4, $5, $6
+                FROM
+                    devices
+                    INNER JOIN accounts ON accounts.id = devices.owner
+                WHERE
+                    accounts.aci = $1
+                    AND accounts.pni = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:494]: Add COALESCE

File: server/src/postgres.rs [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+            WHERE id IN (
+                SELECT
+                    one_time_pre_key_store.id
+                FROM
+                    one_time_pre_key_store
+                    INNER JOIN devices ON devices.id = one_time_pre_key_store.owner
+                    INNER JOIN accounts ON accounts.id = devices.owner
+                WHERE
+                    accounts.aci = $1
+                    AND accounts.pni = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:534]: Add COALESCE

File: server/src/postgres.rs [hunk 14]
```diff
@@ -0,0 +1,586 @@
...
+            r#"
+        INSERT INTO
+            devices (device_id, owner)
+        SELECT
+            $3, id
+        FROM
+            accounts
+        WHERE
+            aci = $1
+            AND pni = $2
```
Comment (Darkros1245) [inline server/src/postgres.rs:571]: Add COALESCE

File: server/src/server.rs [hunk 15]
```diff
@@ -45,244 +49,23 @@ impl PublicKeyType {
...
-    usr_id: UserID,
-    device_id: DeviceID,
+async fn handle_post_keycheck<T: SignalDatabase>(
+    database: T,
+    usr_id: ServiceId,
+    device_id: DeviceId,
     usr_digest: [u8; 32],
 ) -> Result<bool> {
+    todo!()
+    /*
```
Comment (Darkros1245) [inline server/src/server.rs:62]: Add or remove comment

File: server/src/server.rs [hunk 16]
```diff
@@ -426,13 +344,17 @@ pub async fn start_server() -> Result<(), Box<dyn std::error::Error>> {
...
 mod server_tests {
     use super::*;
+    use super::{handle_put_messages, SignalServerState};
+    use crate::database::SignalDatabase;
     use libsignal_protocol::*;
     use rand::rngs::OsRng;
     use sha2::{Digest, Sha256};
+    use uuid::Uuid;
 
+    /*
```
Comment (Darkros1245) [inline server/src/server.rs:354]: Add or remove comment

**Review by Darkros1245 — CHANGES_REQUESTED** [review]

File: server/database/init.sql [hunk 17]
```diff
@@ -84,7 +84,7 @@ VALUES
 INSERT INTO
     devices (device_id, owner)
 SELECT
-    'bob_device', id
```
Comment (UnluckyBird) [inline server/database/init.sql:87]: Mener at deviceId starter fra 1, hvor 1 betyder det er primary device for brugeren

File: server/src/api_error.rs [hunk 18]
```diff
@@ -0,0 +1,19 @@
+use axum::{
+    http::{header, Response, StatusCode},
+    response::IntoResponse,
+    Json,
+};
+use serde_json::json;
+
+pub struct ApiError {
+    pub status_code: StatusCode,
+    pub error_code: Option<u32>,
```
Comment (UnluckyBird) [inline server/src/api_error.rs:10]: Kommer vi til at bruge errorcode ud over bare statuscode?

Comment (Laursen79) [reply server/src/api_error.rs:10]: god pointe

File: server/database/init.sql [hunk 19]
```diff
@@ -1,23 +1,23 @@
...
     identity_key    BYTEA NOT NULL
 );
 
-CREATE TABLE devices (
-    id          INTEGER PRIMARY KEY NOT NULL GENERATED ALWAYS AS IDENTITY,
-    device_id   TEXT NOT NULL,
-    owner       INTEGER NOT NULL REFERENCES accounts(id) ON DELETE CASCADE,
-    UNIQUE(device_id, owner)
-);
+-- CREATE TABLE devices (
```
Comment (Mikkel-Boje) [inline server/database/init.sql:10]: Remove

File: server/database/init.sql [hunk 20]
```diff
@@ -1,23 +1,23 @@
...
     auth_token      VARCHAR(255) NOT NULL UNIQUE,
     identity_key    BYTEA NOT NULL
 );
 
-CREATE TABLE devices (
-    id          INTEGER PRIMARY KEY NOT NULL GENERATED ALWAYS AS IDENTITY,
-    device_id   TEXT NOT NULL,
-    owner       INTEGER NOT NULL REFERENCES accounts(id) ON DELETE CASCADE,
-    UNIQUE(device_id, owner)
-);
```
Comment (Darkros1245) [inline server/database/init.sql:15]: Remove

File: server/database/init.sql [hunk 21]
```diff
@@ -73,30 +75,30 @@ CREATE TABLE one_time_pre_key_store (
...
-INSERT INTO
-    devices (device_id, owner)
-SELECT
-    'bob_device', id
-FROM
-    accounts
-WHERE
-    aci = 'bob'
-    AND pni IS NULL;
+-- INSERT INTO
```
Comment (Mikkel-Boje) [inline server/database/init.sql:86]: Remove

Comment (Darkros1245) [inline server/database/init.sql:86]: Remove

File: server/database/init.sql [hunk 22]
```diff
@@ -73,30 +75,30 @@ CREATE TABLE one_time_pre_key_store (
...
 
-INSERT INTO
-    devices (device_id, owner)
-SELECT
-    'alice_device', id
-FROM
-    accounts
-WHERE
-    aci = 'alice'
-    AND pni IS NULL;
```
Comment (Darkros1245) [inline server/database/init.sql:102]: Remove

File: server/src/postgres.rs [hunk 23]
```diff
@@ -0,0 +1,464 @@
+use crate::{account::Account, database::SignalDatabase};
```
Comment (Darkros1245) [inline server/src/postgres.rs:1]: A genius must have made this code <3

File: server/src/server.rs [hunk 24]
```diff
@@ -45,244 +50,91 @@ impl PublicKeyType {
...
-    database: Arc<Mutex<InMemorySignalDatabase>>,
-    id: UserID,
-    device: DeviceID,
-) -> Result<(), ErrorMessage> {
-    todo!()
+async fn handle_get_messages<T: SignalDatabase>(
+    state: SignalServerState<T>,
+    address: ProtocolAddress,
+) {
+    println!("Get messages")
```
Comment (Darkros1245) [inline server/src/server.rs:110]: Nice handling

File: server/src/database.rs [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    ) -> Result<()>;
+
+    /// Get the keys that are needed to start a conversation with the device that
+    /// corrosponds to the given [ProtocolAddress].
+    async fn get_key_bundle(&self, address: ProtocolAddress) -> Result<DevicePreKeyBundle>;
+
+    /// Get how many keys are left until a last resort key is used instead of
+    /// a one time prekey. More keys should be uploaded when this value is below
+    /// some threshold.
+    async fn get_one_time_pre_key_count(&self, account: &ServiceId) -> Result<u32>;
```
Comment (Mikkel-Boje) [inline server/src/database.rs:55]:
```suggestion
    async fn get_one_time_pre_key_count(&self, service_id: &ServiceId) -> Result<u32>;
```

File: server/src/server.rs [hunk 26]
```diff
@@ -45,244 +50,91 @@ impl PublicKeyType {
...
-    device_id: DeviceID,
+async fn handle_post_keycheck<T: SignalDatabase>(
+    database: T,
+    usr_id: ServiceId,
+    device_id: DeviceId,
     usr_digest: [u8; 32],
 ) -> Result<bool> {
+    todo!()
+    /*
     if let Some(keys) = database
```
Comment (Darkros1245) [inline server/src/server.rs:132]: Remove

File: server/src/database.rs [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+
+    /// Store new one time prekeys to avoid running out.
+    async fn store_one_time_pre_keys(
+        &self,
+        otpks: Vec<UploadSignedPreKey>,
+        owner: ProtocolAddress,
```
Comment (Mikkel-Boje) [inline server/src/database.rs:61]:
```suggestion
        adress: ProtocolAddress,
```

File: server/src/database.rs [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+    ) -> Result<()>;
+
+    /// Get a one time prekey so that you can start a conversation with the
+    /// device that is associated with the given [ProtocolAddress].
+    async fn get_one_time_pre_key(&self, owner: ProtocolAddress) -> Result<UploadSignedPreKey>;
```
Comment (Mikkel-Boje) [inline server/src/database.rs:66]:
```suggestion
    async fn get_one_time_pre_key(&self, adress: ProtocolAddress) -> Result<UploadSignedPreKey>;
```

File: server/src/server.rs [hunk 29]
```diff
@@ -406,35 +277,196 @@ pub async fn start_server() -> Result<(), Box<dyn std::error::Error>> {
...
+    async fn handle_register_account_registers_account() {
+        todo!()
+    }
 
+    #[ignore = "Not implemented"]
     #[tokio::test]
-    async fn keys_check_test() {
+    async fn handle_get_keys_gets_keys() {
+        todo!()
+        /*
```
Comment (Darkros1245) [inline server/src/server.rs:408]: Remove

File: server/src/server.rs [hunk 30]
```diff
@@ -406,35 +277,196 @@ pub async fn start_server() -> Result<(), Box<dyn std::error::Error>> {
...
+            j
+        );
+         */
+    }
+
+    #[ignore = "Not implemented"]
+    #[tokio::test]
+    async fn handle_post_keycheck_test() {
+        todo!()
+        /*
```
Comment (Darkros1245) [inline server/src/server.rs:466]: Remove

File: server/src/in_memory_db.rs [hunk 31]
```diff
@@ -1,17 +1,185 @@
...
+            Err("Device could not be created because user did not exist".into())
+        }
+    }
+}
+*/
+
+#[async_trait]
+impl SignalDatabase for InMemorySignalDatabase {
+    async fn add_account(&self, account: Account) -> Result<()> {
+        self.devices
```
Comment (UnluckyBird) [inline server/src/in_memory_db.rs:89]: Skal nokk være self.accounts, når det er accounts vi adder? Og hvorfor asserter vi?

**Review by Darkros1245 — CHANGES_REQUESTED** [review]

File: server/src/postgres.rs [hunk 32]
```diff
@@ -0,0 +1,464 @@
...
+}
+
+fn parse_service_id(service_id: &ServiceId) -> (Option<String>, Option<String>) {
+    match service_id {
+        ServiceId::Aci(_) => (Some(service_id.service_id_string()), None),
+        ServiceId::Pni(_) => (None, Some(service_id.service_id_string())),
+    }
+}
+
+fn pasrse_to_specific_service_id(service_id: &ServiceId) -> (String, String) {
```
Comment (UnluckyBird) [inline server/src/postgres.rs:448]: parse*

File: server/src/postgres.rs [hunk 33]
```diff
@@ -0,0 +1,464 @@
...
+    match parse_service_id(service_id) {
+        (None, Some(id)) => ("pni".into(), id),
+        (Some(id), None) => ("aci".into(), id),
+        _ => panic!("NOT POSSINLBE!"),
+    }
+}
+
+fn pasrse_to_specific_service_id_from_protocol_address(
```
Comment (UnluckyBird) [inline server/src/postgres.rs:456]: parse*

File: server/src/postgres.rs [hunk 34]
```diff
@@ -0,0 +1,464 @@
...
+            UPDATE
+                accounts
+            SET
+                aci = $3
+            WHERE
+                $1 = $2
+            "#,
+            id_str,
+            id,
+            Some(new_aci.service_id_string())
```
Comment (Mikkel-Boje) [inline server/src/postgres.rs:95]: Please remove Some as it adds nothing
```suggestion
            new_aci.service_id_string()
```

File: server/src/postgres.rs [hunk 35]
```diff
@@ -0,0 +1,464 @@
...
+        UPDATE
+            accounts
+        SET
+            pni = $3
+        WHERE
+            $1 = $2
+        "#,
+            id_str,
+            id,
+            Some(new_pni.service_id_string())
```
Comment (Mikkel-Boje) [inline server/src/postgres.rs:117]:
```suggestion
            new_pni.service_id_string()
```

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]

File: server/src/server.rs [hunk 36]
```diff
@@ -1,27 +1,32 @@
+use crate::api_error::ApiError;
+use crate::database::SignalDatabase;
+use crate::in_memory_db::InMemorySignalDatabase;
+use crate::postgres::PostgresDatabase;
 use anyhow::{bail, Result};
-use axum::extract::{Path, State};
+use axum::extract::{FromRef, FromRequestParts, Path, State};
 use axum::http::header::{ACCEPT, AUTHORIZATION, CONTENT_LENGTH, CONTENT_TYPE, ORIGIN};
```
Comment (UnluckyBird) [inline server/src/server.rs:7]: Er det ikke muligt at neste det her use hell lidt, eller bliver fmt sur så?

File: server/src/server.rs [hunk 37]
```diff
@@ -45,244 +50,91 @@ impl PublicKeyType {
...
-        .usernames
-        .iter()
-        .find(|(_, existing)| **existing == *username)
+async fn handle_put_messages<T: SignalDatabase>(
+    state: SignalServerState<T>,
+    address: ProtocolAddress,
+    payload: Envelope,
+) -> Result<(), ApiError> {
+    println!("Received message");
+    if let Err(result) = state
```
Comment (UnluckyBird) [inline server/src/server.rs:92]:
```
    state.database()
        .push_message_queue(address, vec![payload])
        .await
        .map_err(|_| Err(ApiError {
            message: "Could not push the message to message queue.".to_owned(),
            status_code: StatusCode::INTERNAL_SERVER_ERROR,
        }))?
    OK()
```

**Review by UnluckyBird — CHANGES_REQUESTED** [review]

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]

File: server/src/database.rs (hunk 27 above)
Comment (UnluckyBird) [reply server/src/database.rs:61]: address*

File: server/src/database.rs (hunk 28 above)
Comment (UnluckyBird) [reply server/src/database.rs:66]: address

File: server/src/in_memory_db.rs [hunk 38]
```diff
@@ -1,17 +1,183 @@
...
+            .entry(service_id)
+            .or_insert_with(Vec::new);
+        Ok(())
+    }
+    async fn get_account(&self, service_id: &ServiceId) -> Result<Account> {
+        self.accounts
+            .lock()
+            .await
+            .get(&service_id)
+            .ok_or_else(|| anyhow!("No user exists with ID"))
```
Comment (Mikkel-Boje) [inline server/src/in_memory_db.rs:104]:
```suggestion
            .ok_or(anyhow!("No user exists with ID"))
```

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]

**Review by UnluckyBird — APPROVED** [review]
For sure 100% bug free

**Review by Darkros1245 — APPROVED** [review]

**Review by Mikkel-Boje — APPROVED** [review]
Such a stable api, unlucky that the database is so

**Pull request merged**